fletcher/algorithms/string.py

```diff
@@ -267,6 +268,118 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
     )
 
 
+@njit
+def _get_zstring(width: int, buffer: np.ndarray) -> np.ndarray:
+    num_zeros = width
+    for i in np.bitwise_xor(np.right_shift(buffer, 6), 2):
+        if i != 0:
+            num_zeros -= 1
+    if num_zeros <= 0:
+        return buffer
+    # Create array of byte-representation of 0s concat'd with buffer
+    return np.concatenate((np.full(num_zeros, 48, dtype=np.uint8), buffer), axis=0)
+
+
+@njit
+def _get_zstring_2(
+    width: int, buffer: np.ndarray, temp_buffer: np.ndarray
+) -> np.ndarray:
+    num_zeros = width
+    for i in np.bitwise_xor(np.right_shift(buffer, 6), 2):
+        if i != 0:
+            num_zeros -= 1
+    if num_zeros <= 0:
+        return buffer
+    # Create array of byte-representation of 0s concat'd with buffer
+    for buff_idx in range(num_zeros):
+        temp_buffer[buff_idx] = 48
+    for buff_idx in range(num_zeros, num_zeros + len(buffer)):
+        temp_buffer[buff_idx] = buffer[buff_idx - num_zeros]
+    return temp_buffer[0 : buff_idx + 1]
```

Comment:
Hmm..our several (very slightly varied, one below, another that directly the new values without looping) attempts at creating a local buffer that is reused doesn't seem to bring any performance speedups, which leads me to believe we are doing something dumb. Do any tips come to mind? Wondering if it's still a problem to return the slice as opposed to dealing with this differently. @xhochy

Reply:
I have checked out your code locally and for me `benchmarks.TimeSuite.time_zfill_ext` takes 16ms. Out of these 16ms, 12ms are spent on `str_builder.append_value` and `str_builder.append_null`. I sadly don't see any easy chance to speed this up. It seems like they occur the usual `jitclass` slowdown I have talked about in the beginning. Let's see whether #160 improves that a bit.

Reply:
Bummer but sounds good, thanks. We will then focus more on cleanup + coverage for now then.

Reply:
Current (not final) benchmarks for `pandas` vs `fletcher`:

`[ 75.00%] ··· benchmarks.TimeSuite.time_zfill      10.1±0.1ms`
`[100.00%] ··· benchmarks.TimeSuite.time_zfill_ext  14.9±0.1ms`